tests/conftest.py

```diff
@@ -1,18 +1,42 @@
+import atexit
 import pytest
 import pytest_asyncio
 import tempfile
 from pathlib import Path
+from chia.consensus.constants import ConsensusConstants
+from tests.block_tools import BlockTools, test_constants, create_block_tools
+from tests.util.keyring import TempKeyring
+from typing import Tuple
 
 
-# TODO: tests.setup_nodes (which is also imported by tests.util.blockchain) creates a
-#       global BlockTools at tests.setup_nodes.bt.  This results in an attempt to create
-#       the chia root directory which the build scripts symlink to a sometimes-not-there
-#       directory.  When not there Python complains since, well, the symlink is a file
-#       not a directory and also not pointing to a directory.  In those same cases,
-#       these fixtures are not used.  It would be good to refactor that global state
-#       creation, including the filesystem modification, away from the import but
-#       that seems like a separate step and until then locating the imports in the
-#       fixtures avoids the issue.
+def cleanup_keyring(keyring: TempKeyring):
+    keyring.cleanup()
+
+
+temp_keyring = TempKeyring(populate=True)
+keychain = temp_keyring.get_keychain()
+atexit.register(cleanup_keyring, temp_keyring)  # Attempt to clean up the temp keychain
+
+
+@pytest.fixture(scope="session", name="bt")
+def bt() -> BlockTools:
```

Comment:
I think you will need to make this a generator fixture, otherwise you don't get an opportunity to clean it up (which is critical).

Also, one of the problems with the current global variable `bt` is that it's mutable and tests *do* mutate it. Giving this "session" scope, preserves some of those problems (it's still mutable, right?). Would it be feasible to at least make it "module" scoped?

Reply:
Thanks, I knew I'd need that once I added the keychain fixture. Also, I looked, but I'm not aware of any BlockTools cleanup functions. Let me look at that again.